src/main/java/edu/kit/compiler/optimizations/inlining/InliningOptimization.java

```diff
@@ -0,0 +1,169 @@
+            @Override
+            public void visit(Call call) {
+                var entry  = getCalleeEntry(call);
+                var callee = getEntity(call).getGraph();
+                if (entry.isPresent() && Inliner.canBeInlined(call.getGraph(), callee)) {
+                    double prio = calculatePriority(call, entry.get());
+                    if (entry.get().isAlwaysInline()) {
+                        alwaysInlineCalls.add(call);
+                    } else if (prio >= 0) {
+                        maybeInlineCalls.add(new PrioritizedCall(call, prio));
+                    }
+                }
+            }
+        });
+        maybeInlineCalls.sort(Comparator.reverseOrder());
+    }
+
+    /**
+     * High priority is considered first.
+     *
+     * Returns -1 if the call shouldn't be inlined at all.
+     */
+    private static double calculatePriority(Call call, InliningStateTracker.CalleeEntry entry) {
+        int numConstArgs = 0;
+        for (int i = 2; i < call.getPredCount(); i++) {
+            if (call.getPred(i).getOpCode() == binding_irnode.ir_opcode.iro_Const) {
+                numConstArgs++;
+            }
+        }
+        int logWeight = argWeighting(call.getPredCount() - 2, entry.getNumUsedArgs(), numConstArgs);
+        if (entry.isRecursive()) {
+            // inlining of recursive functions is usually a bad idea
+            logWeight -= 3;
+        }
+        boolean doInline = logWeight >= 3 || (
+                (Math.pow(2, logWeight) * InliningStateTracker.UNPROBLEMATIC_SIZE_INCREASE / 2) >= entry.getNumNodes()
+                        && logWeight >= 0
+        );
+        if (doInline) {
+            double basePrio = Math.pow(2, logWeight) / entry.getNumNodes();
+            return basePrio;
+        }
+        return -1;
+    }
+
+    private static int argWeighting(int nArgs, int nUsedArgs, int nConstArgs) {
+        int weight = nUsedArgs < nArgs ? 1 : 0;
+        if (nConstArgs > 0) {
+            weight += 2;
+            weight += nConstArgs;
+            weight = Math.max(0, weight - nArgs / 5);
+            if (nConstArgs >= nUsedArgs - 1) {
+                weight += 1;
+            }
+            if (nConstArgs == nUsedArgs) {
+                weight += 1;
+            }
+        }
+        return weight;
+    }
+
+    private static Entity getEntity(Call call) {
+        var addr = (Address) call.getPtr();
+        return addr.getEntity();
+    }
+
+    private Optional<InliningStateTracker.CalleeEntry> getCalleeEntry(Call call) {
+        return stateTracker.getCalleeEntry(getEntity(call));
+    }
+
+    @RequiredArgsConstructor
+    private static class PrioritizedCall implements Comparable<PrioritizedCall> {
+        @Getter
+        private final Call call;
+        private final double priority;
+
+
+        @Override
+        public int compareTo(PrioritizedCall o) {
+            return Comparator.<PrioritizedCall>comparingDouble(pc -> pc.priority).compare(this, o);
```

Comment:
It may be worth keeping the `Comparator` as a constant, especially if this is called frequently.